### ISSUE TITLE
GetOrAddCol(colIndex) is actually just Get(colIndex)

### ISSUE THREAD

**Issue by Mertsch** [nietras/Sep#70]
@nietras Is this intended?
```
public Col this[int colIndex] => new Col(_writer.GetOrAddCol(colIndex));
public Col this[string colName] => new Col(_writer.GetOrAddCol(colName));
```
vs.
```
internal ColImpl GetOrAddCol(int colIndex)
{
	return _cols[colIndex]; // This does NOT check the length
}

internal ColImpl GetOrAddCol(string colName)
{
	if ((uint)_cacheIndex < (uint)_colNameCache.Count)
	{
		// [...] This code does
	}
	return value;
}
```

**Comment by nietras** [comment]
I am not exactly sure what you think is the issue here? Can you write
example code that shows unexpected behavior for this? Consider indexing is
bounds checked.

**Comment by Mertsch** [comment]
@nietras As I noticed later on there seems to be a fundamental difference between what I expected and how Sep works.
Consider the following code
```
public static string GetString(params string?[] columns)
{
    if (columns.IsNullOrEmpty())
    {
        return string.Empty;
    }

    using (SepWriter sepWriter = csvWriterOptions.ToText())
    {
        using (SepWriter.Row row = sepWriter.NewRow())
        {
            for (int index = 0; index < columns.Length; index++)
            {
                //row[index].Set(columns[index]); // This code fail, because it expectes already added columns
                row[index.ToString(CultureInfo.InvariantCulture)].Set(columns[index]); // This code adds a new column (with header)
            }
        }
        sepWriter.Flush();
        return sepWriter.ToString();
    }
}
```
This leads to
`["Value1", "Value2", "Value3"]` ->
```
0,1,2
Value1,Value2,Value3
```
What I expected was header-less, single line experience
`["Value1", "Value2", "Value3"]` -> `Value1,Value2,Value3`

Header-less means just **adding** columns one by one and don't care about indices.
Ideally I would write `row.AddCol().Set(columns[index]);` and it would just add a column for index `0` (unnamed) and so on.

Also when reading I noticed that Sep has a very strong focus on headers. When reading multiple line, the first line basically locks down the remaining shape of the CSV.
In my expectation I would be able to read just line by line and don't care about the amount of columns, as I do not have headers anyway.

So my question would be will there be a version of Sep that is less header focused?!

I did some benchmarks and the numbers are insane!
```
| Method              | Mean        | Error    | StdDev   | Ratio | Gen0   | Gen1   | Allocated | Alloc Ratio |
|-------------------- |------------:|---------:|---------:|------:|-------:|-------:|----------:|------------:|
| GetEntriesCsvHelper | 10,661.3 ns | 19.47 ns | 16.26 ns |  1.00 | 1.5259 | 0.1221 |   25600 B |        1.00 |
| GetEntriesSep       |    264.9 ns |  2.18 ns |  1.82 ns |  0.02 | 0.0501 |      - |     840 B |        0.03 |

| Method                  | Mean        | Error     | StdDev    | Median      | Ratio | Gen0   | Gen1   | Allocated | Alloc Ratio |
|------------------------ |------------:|----------:|----------:|------------:|------:|-------:|-------:|----------:|------------:|
| GetRowsAndColsCsvHelper | 23,330.7 ns | 457.10 ns | 670.02 ns | 22,905.1 ns |  1.00 | 3.0518 | 0.2441 |   50.3 KB |        1.00 |
| GetRowsAndColsSep       |    851.5 ns |  16.40 ns |  16.11 ns |    841.7 ns |  0.04 | 0.1326 |      - |   2.18 KB |        0.04 |

| Method             | Mean       | Error    | StdDev   | Ratio | Gen0   | Gen1   | Allocated | Alloc Ratio |
|------------------- |-----------:|---------:|---------:|------:|-------:|-------:|----------:|------------:|
| GetStringCsvHelper | 1,606.1 ns | 31.59 ns | 60.10 ns |  1.00 | 0.8984 | 0.0172 |  14.67 KB |        1.00 |
| GetStringSep       |   311.7 ns |  6.26 ns |  8.77 ns |  0.20 | 0.1192 | 0.0005 |   1.95 KB |        0.13 |
```
Based on these numbers I have a strong urge to switch to Sep, but currently I can only adopt it for single line reading of CSV, since multi-line reading and writing in general are totally header focussed.

Here are a few test cases I am testing against
```
[TestClass]
public sealed class CsvStringHelperTests
{
    [TestMethod]
    [DataRow(null, true, "")]
    [DataRow(null, false, "")]
    [DataRow(@"", true, "")]
    [DataRow(@"", false, "")]
    [DataRow(@" x ", true, "x")]
    [DataRow(@" x ", false, " x ")]
    [DataRow(@"x,y", true, "x;y")]
    [DataRow(@"x,y", false, "x;y")]
    [DataRow(@" x , y ", true, "x;y")]
    [DataRow(@" x , y ", false, " x ; y ")]
    [DataRow("""1,"x","y",1""", true, "1;x;y;1")]
    [DataRow("""2,"x","y",2""", false, "2;x;y;2")]
    [DataRow("""3," x ", " y " ,3""", true, """3;x;" y ";3""")]
    [DataRow("""4," x ", " y " ,4""", false, """4; x ; " y " ;4""")]
    [DataRow("""5,a, b,c , d ,"e"," f","g "," h ", "i","j" , "k" , " l "," m " , " n " ,5""", true,
        """5;a;b;c;d;e;f;g;h;"i";j;"k";" l ";m;" n ";5""")]
    [DataRow("""5,a, b,c , d ,"e"," f","g "," h ", "i","j" , "k" , " l "," m " , " n " ,5""", false,
        """5;a; b;c ; d ;e; f;g ; h ; "i";j ; "k" ; " l "; m  ; " n " ;5""")]
    public void GatherFilesTest(string csvString, bool trim, string expectedResult)
    {
        CsvStringHelperSep.GetEntries(csvString, trim).JoinToString(";").ShouldBe(expectedResult);

        CsvStringHelper.GetEntries(csvString, trim).JoinToString(";").ShouldBe(expectedResult switch
            {
                "3;x;\" y \";3" => "3;x",
                "4; x ; \" y \" ;4" => "4; x ",
                "5;a; b;c ; d ;e; f;g ; h ; \"i\";j ; \"k\" ; \" l \"; m  ; \" n \" ;5" => "5;a; b;c ; d ;e; f;g ; h ",
                "5;a;b;c;d;e;f;g;h;\"i\";j;\"k\";\" l \";m;\" n \";5" => "5;a;b;c;d;e;f;g;h",
                _ => expectedResult,
            });
    }

    [TestMethod]
    [DataRow(null, true, "")]
    [DataRow(null, false, "")]
    [DataRow(@"", true, "")]
    [DataRow(@"", false, "")]
    [DataRow(@" x ", true, "x")]
    [DataRow(@" x ", false, " x ")]
    [DataRow(@"x,y", true, "x;y")]
    [DataRow(@"x,y", false, "x;y")]
    [DataRow(@" x , y ", true, "x;y")]
    [DataRow(@" x , y ", false, " x ; y ")]
    [DataRow(" x \n a ", true, "x|a")]
    [DataRow(" x \n a ", false, " x | a ")]
    [DataRow("x,y\na,b", true, "x;y|a;b")]
    [DataRow("x,y\na,b", false, "x;y|a;b")]
    [DataRow(" x , y \n a , b ", true, "x;y|a;b")]
    [DataRow(" x , y \n a , b ", false, " x ; y | a ; b ")]
    [DataRow("x\na,b\n1,2,3,4", false, "x|a;b|1;2;3;4")]
    public void GetRowsAndColsTest(string csvString, bool trim, string expectedResult)
    {
        if (csvString is "x\na,b\n1,2,3,4")
        {
            Assert.ThrowsException<InvalidDataException>(() => CsvStringHelperSep.GetRowsAndCols(csvString, trim).ToArray())
                .Message.ShouldStartWith("Found 2 column(s) on row 1/lines [2..3]:'a'");
        }
        else
        {
            CsvStringHelperSep.GetRowsAndCols(csvString, trim).Select(f => f.JoinToString(";")).JoinToString("|").ShouldBe(expectedResult);
        }
        CsvStringHelper.GetRowsAndCols(csvString, trim).Select(f => f.JoinToString(";")).JoinToString("|").ShouldBe(expectedResult);
    }

    [TestMethod]
    public void GetStringTest()
    {
        CsvStringHelperSep.GetString(null!).ShouldBe(string.Empty);
        CsvStringHelperSep.GetString([]).ShouldBe(string.Empty);
        CsvStringHelperSep.GetString([null]).ShouldBe("0", "Should actually be ''");
        CsvStringHelperSep.GetString(string.Empty).ShouldBe("0", "Should actually be ''");
        CsvStringHelperSep.GetString([null, null]).ShouldBe("0,1\r\n,", "Should actually be ','");
        CsvStringHelperSep.GetString([string.Empty, null]).ShouldBe("0,1\r\n,", "Should actually be ','");
        CsvStringHelperSep.GetString([string.Empty, string.Empty]).ShouldBe("0,1\r\n,", "Should actually be ','");
        CsvStringHelperSep.GetString("Hello").ShouldBe("0\r\nHello", "Should actually be 'Hello'");
        CsvStringHelperSep.GetString("Hello", "World").ShouldBe("0,1\r\nHello,World", "Should actually be 'Hello,World'");
        CsvStringHelperSep.GetString("Hello, ", "Special chars like: \", ',\r\n,\n, ;, ., :").ReplaceLineEndings().ShouldBe(
            """
            0,1
            Hello, ,Special chars like: ", ',
            ,
            , ;, ., :
            """,
            """
            Should actually be
            "Hello, ","Special chars like: "", ',
            ,
            , ;, ., :"
            """);

        CsvStringHelper.GetString(null!).ShouldBe(string.Empty);
        CsvStringHelper.GetString([]).ShouldBe(string.Empty);
        CsvStringHelper.GetString([null]).ShouldBe(string.Empty);
        CsvStringHelper.GetString(string.Empty).ShouldBe(string.Empty);
        CsvStringHelper.GetString([null, null]).ShouldBe(",");
        CsvStringHelper.GetString([string.Empty, null]).ShouldBe(",");
        CsvStringHelper.GetString("Hello").ShouldBe("Hello");
        CsvStringHelper.GetString("Hello", "World").ShouldBe("Hello,World");
        CsvStringHelper.GetString("Hello, ", "Special chars like: \", ',\r\n,\n, ;, ., :").ReplaceLineEndings().ShouldBe(
            """
            "Hello, ","Special chars like: "", ',
            ,
            , ;, ., :"
            """);
    }
}
```

As last time (with string unescaping) this is a blocker for switching to Sep for me, so if you say you don't want to go this route then I have to stick with CSVHelper, if not I'd be happy to assist in providing more feedback if necessary 😄 

Thanks for your great work and time!

**Comment by nietras** [comment]
@Mertsch SepWriter does not currently support no header scenarios. Basically because I/we don't need for our usage. I am sorry this is not clear from README, so will have to look at improving that. And yes, Sep is heavily focused on header scenarios because we would never use header less for the work we do, I would never recommend any one anything else, but I understand there are different needs. 

Headerless writing is definitely something I'd like to add it's just not a priority compared to other things currently. 

There is still a lot of things to do on in particular the writer, main focus has been on reader and its perf, because that's what I find fun. 😊

String unescaping is supported now, so not sure why this a blocker anymore?

Thanks 😊

**Comment by Mertsch** [comment]
@nietras I meant after the escaping **was** blocking, this is the new blocking thing 😄 
So I guess you may close the issue and I wait till you suddenly open it again and surprise me 🤣 

Happy christmas 🎄

**Comment by Mertsch** [comment]
@nietras I went a bit crazy ... out of curiosity and ...
This super un-optimized code (string splitting) ignores headers and is still crazy fast.
What do you think?
With a proper SpanReader that just moves forward this could be even faster. Also is there a chance to know the `Current` rows length including the new line character variations?!
```
[MustUseReturnValue]
public static IEnumerable<string[]> GetRowsAndCols(string? input, bool trimEntries = false)
{
    if (input.IsNullOrEmpty())
    {
        yield break;
    }

    int currentPos = 0;
    while (currentPos < input.Length)
    {
        using SepReader csvReader = csvReaderOptions.FromText(input[currentPos..]);
        if (csvReader.MoveNext())
        {
            yield return getCols(csvReader, trimEntries).ToArray();
            currentPos += csvReader.Current.Span.Length;
            if (currentPos<input.Length && input[currentPos] is '\r' or '\n')
            {
                currentPos++;
                if (currentPos < input.Length && input[currentPos] is '\n')
                {
                    currentPos++;
                }
            }
        }
    }
}
```
```
| Method                  | Mean      | Error     | StdDev    | Ratio | RatioSD | Gen0   | Gen1   | Allocated | Alloc Ratio |
|------------------------ |----------:|----------:|----------:|------:|--------:|-------:|-------:|----------:|------------:|
| GetRowsAndColsCsvHelper | 47.222 us | 0.9268 us | 1.7180 us |  1.00 |    0.00 | 3.8452 | 0.2441 |  62.75 KB |        1.00 |
| GetRowsAndColsSep       |  1.618 us | 0.0299 us | 0.0293 us |  0.03 |    0.00 | 0.2804 |      - |   4.61 KB |        0.07 |
```

**Comment by nietras** [comment]
@Mertsch I have added three new issues #71, #72, #73 tracking what I presume are missing features for SepWriter according to most people probably. No idea when I will address these, but hopefully this makes it clear what is missing.

Regarding the code you show this will be incredibly slow for both CsvHelper and Sep compared to straight up parsing, since you are allocating a lot per row/line here, when creating a reader per line. It also will be very bad with upcoming changes to Sep where it will parse multiple rows per MoveNext in batched way. I don't know why you are trying to do it this way, but perhaps you should consider giving `Sylvan.Data.Csv` a try, this is fully featured incl. trimming as far as I know, and also fast. :)

I'm guessing you would also like trim support as far as I can tell so added issue for that #74 too. Again no promises but hopefully this makes it clear.

Thanks again for your feedback!

**Comment by nietras** [comment]
@Mertsch note that all parsers I know will copy data from source (file, stringreader or similar) to an internal buffer no matter what and this would typically be several KBs e.g. 32KB and this will happen whenever you create a reader. Since Sep uses ArrayPool there are no repeated allocs so one of the things you are seeing is simply that CsvHelper is allocating vs Sep isn't. No matter what this way of doing create reader per row will be exceedingly inefficient 😉 So I hope you find another way 😅

**Comment by Mertsch** [comment]
@nietras As said, just out of curiosity 😄.
I have actually tried `Sylvan` just yesterday and noticed that it does not fit my needs at all. Reading instantly fails on non-well-formatted CSV and writing is 20 times slower than CSVHelper.

**Comment by nietras** [comment]
> non-well-formatted CSV

Yeah that will always be an issue, and supporting such is not a priority for Sep. CsvHelper no doubt has much better all-round and battletested support for a lot of edge and corner cases, to do that there is a cost of course, so much harder to be fast given that. 

> writing is 20 times slower than CSVHelper.

That surprises me. You might consider filing an issue for Sylvan to ask author of that whether that seems right.

**Comment by Mertsch** [comment]
@nietras Most likely because the writing is `DbDataReader` based which has a huge overhead for writing just one row.

**Comment by nietras** [comment]
Closing this as replaced by separate issues. I hope to be able to address your issues at some point, so stick around ;)

**Comment by nietras** [comment]
WriteHeader false Merged #71
